processing/src/main/java/org/apache/druid/math/expr/ExprEval.java

```diff
@@ -1314,38 +1337,87 @@ public ExpressionType type()
     @Override
     public boolean isNumericNull()
     {
-      return false;
+      if (ExpressionType.NESTED_DATA.equals(expressionType)) {
+        computeNumber();
+        return number == null;
+      }
+      return true;
     }
 
     @Override
     public int asInt()
     {
+      if (ExpressionType.NESTED_DATA.equals(expressionType)) {
+        computeNumber();
+        if (number != null) {
+          return number.intValue();
+        }
+      }
       return 0;
     }
 
     @Override
     public long asLong()
     {
+      if (ExpressionType.NESTED_DATA.equals(expressionType)) {
+        computeNumber();
+        if (number != null) {
+          return number.longValue();
+        }
+      }
       return 0;
     }
 
     @Override
     public double asDouble()
     {
+      if (ExpressionType.NESTED_DATA.equals(expressionType)) {
+        computeNumber();
+        if (number != null) {
+          return number.doubleValue();
+        }
+      }
       return 0;
     }
 
     @Override
     public boolean asBoolean()
     {
+      if (ExpressionType.NESTED_DATA.equals(expressionType)) {
+        Object val = StructuredData.unwrap(value);
+        if (val != null) {
+          return Evals.objectAsBoolean(val);
+        }
+      }
       return false;
     }
 
+    private void computeNumber()
+    {
+      if (number == null && value != null) {
```

Comment:
Seems like the intent is we cache the number in `number`. If so— that doesn't work in the case where `value != null` but `number` _is_ `null` (like, maybe `val` is a string that is not parseable as a number).